### PR TITLE
add Xingyu's GSoC21 submission

### PR DESCRIPTION
**Pull request by charlotte12l**
Hi @serge-sans-paille @rgommers, could you help review my report for GSoC'21 Final Submission?

>As part of the final evaluation, all participating students must provide a link to the work they have done for the program. Failure to do so properly may result in failing the program. There are several different ways to do this, so please read this document carefully.
>These links will be published on the public project list. They help demonstrate the work that was done during the program. They are also a great way for you to refer back to your work for future employers.
>In brief: The target of the link should contain a short description of what work was done, what code got merged, what code didn't get merged, and what's left to do. The best examples of this we saw in past years were "final reports" that made it easy to find the code, summarized the current state of the project, and enumerated challenges and learnings.

I looked at some previous students' report examples so I made a brief blog to meet the requirements. It is not as interesting as other Pythran stories :)

### CODE REVIEW

File: content/2021-08-19-GSoC21.rst [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+:summary: This project is about using Pythran to accelerate algorithms in SciPy and 
+          writing benchmarks for the algorithms. Let's look into the details of the project.
+
+
+
+Project Overview
+================
+There are a lot of algorithms in scipy that use Cython to improve 
+the performance of code that would be too slow as pure Python, 
+e.g. algorithms in ``scipy.spatial``, ``scipy.stats`` and ``scipy.optimize``, etc.
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: `, etc.` -> `.`
(the `.e.g.` and `etc.` are kind of duplicate now)

File: content/2021-08-19-GSoC21.rst [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+* Find and solve potential issues in Pythran
+
+
+My full proposal can be accessed `here <https://docs.google.com/document/d/1nM7dYbmModiukQw-sSOVGz6t5S6HC0VVWucYadI_aMQ/edit?usp=sharing>`_.
+
+
+What I have done
+================
+
+**Pull Requests**:
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: Do you want to split this list in contributions to SciPy and to Pythran? The latter may be nice to separate out and say a few words about. I think that made your project more interesting and challenging overall, so it'd also be good to show that.

File: content/2021-08-19-GSoC21.rst [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+#. [Merged] `General implementation of supporting immediate arguments <https://github.com/serge-sans-paille/pythran/pull/1878>`_: Generalize the above two solutions to support immediate arguments.
+#. [Under Review] `ENH: Pythran implementation of _compute_prob_outside_square and _compute_prob_inside_method to speedup stats.ks_2samp <https://github.com/scipy/scipy/pull/13957>`_
+#. [Under Review] `BUG: fix stats.binned_statistic_dd issue with values close to bin edge <https://github.com/scipy/scipy/pull/14338>`_
+#. [Under Review] `ENH: improved binned_statistic_dd via Pythran <https://github.com/scipy/scipy/pull/14345>`_ 
+#. [Under Review] `ENH: improve cspline1d, qspline1d, and relative funcs via Pythran <https://github.com/scipy/scipy/pull/14429>`_ 
+#. [Under Review] `ENH: improve siegelslopes via pythran <https://github.com/scipy/scipy/pull/14430>`_ 
+#. [Hold] `ENH: Pythran implementation of _cdf_distance <https://github.com/scipy/scipy/pull/14154>`_ : Pythran version is slightly better than the Python one after fixing ``np.searchsorted()``. When SciPy begin using SIMD in the future, it may be faster so this PR is currently on hold.
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: minor: I'd say "On hold" instead of "Hold".

File: content/2021-08-19-GSoC21.rst [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+#. [Hold] `WIP: ENH: improve _count_paths_outside_method via pythran <https://github.com/scipy/scipy/pull/14314>`_ : This PR got stuck in a Mac specific error and we haven’t find out why.
+#. [Hold] `WIP: ENH: improve sort_vertices_of_regions via Pythran and made it more readable <https://github.com/scipy/scipy/pull/14376>`_ : There are currently two tests we can’t pass because 1. With Pythran we can’t do inplace sort 2. The input type will change in the Pythran function
+#. [Closed] `ENH: improve _sosfilt_float via Pythran <https://github.com/scipy/scipy/pull/14473>`_  : ``_sosfilt_float`` is already implemented in Cython. We were considering to replace it but found Pythran performance is not much better than Cython's, and Pythran does not support ``object`` type, so we decided not to merge it.
+
+Since one of our goals is to find potential issues in Pythran and make Pythran better, I also bring up many issues in Pythran.
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: "bring" -> "brought"

File: content/2021-08-19-GSoC21.rst [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+#. [closed] `can't use np.expand_dims with specified keyword argument <https://github.com/serge-sans-paille/pythran/issues/1850>`_
+#. [open ] `bus error on Mac but works fine on Linux for _count_paths_outside_method pythran version <https://github.com/scipy/scipy/issues/14315>`_ 
+#. [open ] `array assignment res[cond1] = ax[cond1] works fine for int[] or float[] or float[:,:] but not int[:,:] <https://github.com/serge-sans-paille/pythran/issues/1858>`_
+
+Work Left
+=========
+
+As the project proceeded, I found it was difficult to find 
+suitable algorithms to be implemented. Moreover, in our past experience 
+with Pythran,  we often run into some things that are easy to get wrong.
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: It'd be nice to add what those things were. E.g.: `... to get wrong, such as the use of integer dtypes in a portable way, and using arrays that are views as input to a Pythran-ized function".

File: content/2021-08-19-GSoC21.rst [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+It has been a great experience working on this project in GSoC'21, 
+my mentors are really friendly and responsive, 
+and the community are also always willing to help. 
+
+
+Special thanks to my mentors, Ralf and Serge, who provided immense support 
+for me to get through the difficulties.
+I’m very fortunate to get the chance to dive into and contribute to SciPy 
+and Pythran this summer, especially with such awesome mentors. 
+I have learnt a lot, intellectual or spiritual, and I would love to continue contributing to SciPy and Pythran in the future :)
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: we'd like that as well :)

**Review by rgommers — COMMENTED** [review]
Thanks @charlotte12l! I made a few suggestions about the story line, as well as some minor textual things.

File: content/2021-08-19-GSoC21.rst [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+:lang: en
+:authors: Xingyu Liu
+:summary: This project is about using Pythran to accelerate algorithms in SciPy and 
+          writing benchmarks for the algorithms. Let's look into the details of the project.
+
+
+
+Project Overview
+================
+There are a lot of algorithms in scipy that use Cython to improve
```
Comment (serge-sans-paille) [inline content/2021-08-19-GSoC21.rst]: s/scipy/SciPy/ for consistency

File: content/2021-08-19-GSoC21.rst [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+
+Project Overview
+================
+There are a lot of algorithms in scipy that use Cython to improve 
+the performance of code that would be too slow as pure Python, 
+e.g. algorithms in ``scipy.spatial``, ``scipy.stats`` and ``scipy.optimize``, etc. 
+Recently, SciPy added experimental support for Pythran, 
+to make it easier to accelerate Python code. 
+Compared with Cython, Pythran is more readable and even faster.
```
Comment (serge-sans-paille) [inline content/2021-08-19-GSoC21.rst:20]: It's probably a good think to add reference to the SciPy, Cython and Pythran website for the first occurence

File: content/2021-08-19-GSoC21.rst [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+* Writing benchmarks for the algorithms in SciPy
+* Find and solve potential issues in Pythran
+
+
+My full proposal can be accessed `here <https://docs.google.com/document/d/1nM7dYbmModiukQw-sSOVGz6t5S6HC0VVWucYadI_aMQ/edit?usp=sharing>`_.
+
+
+What I have done
+================
+
```
Comment (serge-sans-paille) [inline content/2021-08-19-GSoC21.rst:35]: It would be great to structure these PR by sub-themes

File: content/2021-08-19-GSoC21.rst [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+with Pythran,  we often run into some things that are easy to get wrong. 
+Therefore, we need better testing  and we decided to change the plan to 
+write better testing infrastructure for Pythran extensions: 
+`WIP: TST: add tests for Pythran somersd <https://github.com/scipy/scipy/pull/14559#>`_
+
+
+Project Experience
+==================
+It has been a great experience working on this project in GSoC'21, 
+my mentors are really friendly and responsive,
```
Comment (serge-sans-paille) [inline content/2021-08-19-GSoC21.rst:125]: (joke inside) you should expand on that :-)

**Comment by charlotte12l** [comment]
Thank you so much for your comments! I made revisions and could you please check the current version again?

File: content/2021-08-19-GSoC21.rst [hunk 11]
```diff
@@ -33,31 +33,58 @@ My full proposal can be accessed `here <https://docs.google.com/document/d/1nM7d
...
+#. |SciPy| |accelerate| [On Hold] `WIP: ENH: improve _count_paths_outside_method via pythran <https://github.com/scipy/scipy/pull/14314>`_ : This PR got stuck in a Mac specific error and we haven’t find out why.
+#. |SciPy| |accelerate| [On Hold] `WIP: ENH: improve sort_vertices_of_regions via Pythran and made it more readable <https://github.com/scipy/scipy/pull/14376>`_ : There are currently two tests we can’t pass because 1. With Pythran we can’t do inplace sort 2. The input type will change in the Pythran function
+#. |SciPy| |accelerate| [Closed] `ENH: improve _sosfilt_float via Pythran <https://github.com/scipy/scipy/pull/14473>`_  : ``_sosfilt_float`` is already implemented in Cython. We were considering to replace it but found Pythran performance is not much better than Cython's, and Pythran does not support ``object`` type, so we decided not to merge it.
+
+
+**Pythran**
+
+When using Pythran to improve SciPy algorithms, I found some important modules are not 
+supported or got false result in Pythran currently, e.g. boolean arguments 
+such as ``keepdims`` was not supported in Pythran because the return type
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: minor textual: `was` -> `were`

File: content/2021-08-19-GSoC21.rst [hunk 12]
```diff
@@ -33,31 +33,58 @@ My full proposal can be accessed `here <https://docs.google.com/document/d/1nM7d
...
+would change based on true/false of ``keepdims``. Therefore, I made a general
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: `... on true/false of keepdims` ->
```
... on the value of ``keepdims`` (True or False).
```

File: content/2021-08-19-GSoC21.rst [hunk 13]
```diff
@@ -33,31 +33,58 @@ My full proposal can be accessed `here <https://docs.google.com/document/d/1nM7d
...
+#. |Pythran| |feature| [Merged] `Import test cases from scipy <https://github.com/serge-sans-paille/pythran/pull/1830>`_ : Import Pythran functions in SciPy as test case in Pythran
+#. |Pythran| |feature| [Merged] `Feature/add keep dims <https://github.com/serge-sans-paille/pythran/pull/1869#>`_ : support keepdims argument in ``np.mean()`` in Pythran
+#. |Pythran| |feature| [Merged] `Support boolean arguments in numpy unique <https://github.com/serge-sans-paille/pythran/pull/1876>`_
+#. |Pythran| |feature| [Merged] `General implementation of supporting immediate arguments <https://github.com/serge-sans-paille/pythran/pull/1878>`_: Generalize the above two solutions to support immediate arguments.
+
+
+Issues
+------
+
+In addition to the above mentioned, I digged out more issues in Pythran while
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: `above mentioned` --> `above-mentioned issues`

Comment (rgommers) [reply content/2021-08-19-GSoC21.rst]: `digged out` -> `dug up`

File: content/2021-08-19-GSoC21.rst [hunk 14]
```diff
@@ -33,31 +33,58 @@ My full proposal can be accessed `here <https://docs.google.com/document/d/1nM7d
...
+using it, so I opened many issues in Pythran. My mentors often helped solve 
+those issues and I'll test if it works.
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: ` and I'll test if it works. ` -> `and then I tested whether the fixes worked`.

File: content/2021-08-19-GSoC21.rst [hunk 15]
```diff
@@ -70,16 +97,24 @@ Since one of our goals is to find potential issues in Pythran and make Pythran b
...
+#. [open] `array assignment res[cond1] = ax[cond1] works fine for int[] or float[] or float[:,:] but not int[:,:] <https://github.com/serge-sans-paille/pythran/issues/1858>`_
 
 Work Left
 =========
 
 As the project proceeded, I found it was difficult to find 
-suitable algorithms to be implemented. Moreover, in our past experience 
-with Pythran,  we often run into some things that are easy to get wrong. 
-Therefore, we need better testing  and we decided to change the plan to 
+suitable algorithms to be implemented. A suitable algorithm for Pythran should meet  at least three requirements:
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: minor: `meet  at` has a double space in it

File: content/2021-08-19-GSoC21.rst [hunk 16]
```diff
@@ -70,16 +97,24 @@ Since one of our goals is to find potential issues in Pythran and make Pythran b
...
+
+* Is currently slow
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: I'd suggest:
```
* It is currently slow
* It does not have ...
* It has ...
* ```

File: content/2021-08-19-GSoC21.rst [hunk 17]
```diff
@@ -70,16 +97,24 @@ Since one of our goals is to find potential issues in Pythran and make Pythran b
...
-Therefore, we need better testing  and we decided to change the plan to 
+suitable algorithms to be implemented. A suitable algorithm for Pythran should meet  at least three requirements:
+
+* Is currently slow 
+* Do not have modules that Pythran doesn't support, e.g. class type, imported SciPy modules.
+* Has obvious loops so that the speedup would be large. 
+
+I looked through almost all the algorithms but found little.
+Moreover, in our past experience 
+with Pythran,  we often run into some things that are easy to get wrong, such as
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: double space in `,  we`

File: content/2021-08-19-GSoC21.rst [hunk 18]
```diff
@@ -70,16 +97,24 @@ Since one of our goals is to find potential issues in Pythran and make Pythran b
...
+suitable algorithms to be implemented. A suitable algorithm for Pythran should meet  at least three requirements:
+
+* Is currently slow 
+* Do not have modules that Pythran doesn't support, e.g. class type, imported SciPy modules.
+* Has obvious loops so that the speedup would be large. 
+
+I looked through almost all the algorithms but found little.
+Moreover, in our past experience 
+with Pythran,  we often run into some things that are easy to get wrong, such as 
+using arrays that are views as input to a Pythranized function, the use of different dtypes.
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: `, the use of` -> `, or the use of`

File: content/2021-08-19-GSoC21.rst [hunk 19]
```diff
@@ -95,7 +130,14 @@ Special thanks to my mentors, Ralf and Serge, who provided immense support
 for me to get through the difficulties.
 I’m very fortunate to get the chance to dive into and contribute to SciPy 
 and Pythran this summer, especially with such awesome mentors. 
-I have learnt a lot, intellectual or spiritual, and I would love to continue contributing to SciPy and Pythran in the future :)
+I have learnt a lot, intellectual or spiritual. I would love to continue contributing to SciPy and Pythran in the future :)
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: maybe: `intellectual or spiritual` -> `both intellectually and spiritually`

File: content/2021-08-19-GSoC21.rst [hunk 20]
```diff
@@ -95,7 +130,14 @@ Special thanks to my mentors, Ralf and Serge, who provided immense support
...
 
 
 Thanks to Google Summer of Code and Python Software Foundation!
```
Comment (rgommers) [inline content/2021-08-19-GSoC21.rst]: add "the" before Python

**Review by rgommers — COMMENTED** [review]
Starting to look pretty good, only a number of small textual comments from me.

**Comment by charlotte12l** [comment]
Just pushed a second revision :)

**Review by rgommers — COMMENTED** [review]
This is looking pretty good to me now!

One last question: I haven't looked at the rendered site, however the .rst preview shows that `[closed]` isn't nicely rendered like the other tags, is that a problem?

image

**Comment by charlotte12l** [comment]
Oh actually I didn't add image for closed/merged, do you think it is better to also make it a tag?

**Comment by rgommers** [comment]
> Oh actually I didn't add image for closed/merged, do you think it is better to also make it a tag?

Ah I see now, there are badges linked at the bottom of the file. It's a minor thing so if you like it better as it is now then that's fine with me too.

**Review by rgommers — APPROVED** [review]

**Comment by charlotte12l** [comment]
@serge-sans-paille  Hi Serge, if possible, could you please help merge the PR and publish it on the website? I need to provide the blog link to GSoC by 08/23 :)

**Comment by serge-sans-paille** [comment]
http://serge-sans-paille.github.io/pythran-stories/gsoc21-improve-performance-through-the-use-of-pythran.html

**Comment by rgommers** [comment]
Looks great!

**Pull request merged**